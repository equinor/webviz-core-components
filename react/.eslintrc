{
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 6,
        "sourceType": "module"
    },
    "extends": [
        "plugin:react/recommended",
        "eslint:recommended"
    ],
    "rules": {
        "max-len": [
            "error",
            {
                "code": 120
            }
        ]
    },
    "overrides": [
        {
            "files": [
                "*.ts",
                "*.tsx"
            ],
            "plugins": [
                "@typescript-eslint"
            ],
            "extends": [
                "plugin:react/recommended",
                "plugin:@typescript-eslint/recommended"
            ]
        }
    ],
    "settings": {
        "react": {
            "version": "detect"
        }
    },
    "env": {
        "amd": true,
        "browser": true,
        "es6": true,
        "jasmine": true,
        "node": true
    }
}
