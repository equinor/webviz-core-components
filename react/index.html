<!doctype>
<html>
    <head>
        <title>my-dash-component</title>
    </head>
    <body>
        <div id="root"></div>
        <svg width="400px" height="400px" style="margin-top: 50px;">
            <rect width="100px" height="100px" style="fill:green" id="green-rect" />
            <rect width="100px" height="100px" x="200px" style="fill:blue" id="blue-rect" />
        </svg>
        <script src="/output.js"></script>
    </body>
</html>
