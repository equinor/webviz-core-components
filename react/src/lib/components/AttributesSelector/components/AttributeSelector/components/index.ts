export { CollapseExpandButton } from "./CollapseExpandButton";
export { ValuesList } from "./ValuesList";
