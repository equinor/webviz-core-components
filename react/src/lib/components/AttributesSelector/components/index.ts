export { AttributeSelector } from "./AttributeSelector/attribute-selector";
