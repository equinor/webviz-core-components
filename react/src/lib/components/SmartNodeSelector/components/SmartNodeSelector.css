/**
 * Copyright (c) 2021- Equinor ASA
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

.SmartNodeSelector *,
.SmartNodeSelector__Ruler {
    font-family: sans-serif;
}

.SmartNodeSelector--Error {
    background-color: #ffc1c1;
    border: 2px #eb0000 solid;
    border-radius: 4px;
    color: #eb0000;
    padding: 8px;
    font-family: monospace;
    line-height: 1.5em;
}

.SmartNodeSelector__unselectable {
    -webkit-touch-callout: none;
    -webkit-user-select: none;
    -khtml-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

.SmartNodeSelector {
    border: 1px #dcdcdc solid;
    border-radius: 4px;
    padding: 4px;
    padding-left: 8px;
    padding-right: 40px;
    display: flex;
    flex-wrap: wrap;
    cursor: text;
    position: relative;
    margin-top: 8px;
    margin-bottom: 8px;
    min-height: 20px;
    min-width: 200px;
}

.SmartNodeSelector--frameless {
    border: 0;
    padding: 0;
    padding-right: 40px;
}

.SmartNodeSelector--Invalid {
    border: 1px #eb0000 solid !important;
}

.SmartNodeSelector--SuggestionsActive {
    border-radius: 4px 4px 0px 0px;
}

.SmartNodeSelector__ClearAll {
    position: absolute;
    right: 8px;
    top: 50%;
    margin-top: -16px;
}

.SmartNodeSelector__ClearAll button {
    appearance: none;
    background-color: #007079;
    border-radius: 50%;
    border: none;
    cursor: pointer;
    height: 32px;
    margin-left: 8px;
    width: 32px;
    display: flex;
    outline: none;
    position: relative;
    padding: 0;
    background-image: url("./images/x.svg");
    background-size: 18px 18px;
    background-position: center center;
    background-repeat: no-repeat;
}

.SmartNodeSelector__BrowseButtons {
    background: #007079;
    width: 20px;
}

.SmartNodeSelector__ShiftNode {
    appearance: none;
    background: #007079;
    border: none;
    border-radius: 0px;
    cursor: pointer;
    height: 20px;
    margin-right: 8px;
    width: 20px;
    display: inline-block;
    outline: none;
    padding: 0;
    background-size: 16px 16px;
    background-position: center center;
    background-repeat: no-repeat;
    height: 50%;
    position: absolute;
}

.SmartNodeSelector__ShiftUp {
    background-image: url("./images/chevron-up.svg");
    top: 0;
    border-top-left-radius: 2px;
}

.SmartNodeSelector__ShiftDown {
    background-image: url("./images/chevron-down.svg");
    top: 50%;
    border-bottom-left-radius: 2px;
}

.SmartNodeSelector__ShiftNode:disabled,
.SmartNodeSelector__ClearAll button:disabled {
    background-color: #eaeaea;
    cursor: default;
}

.SmartNodeSelector__ShiftNode:hover:enabled,
.SmartNodeSelector__ClearAll button:hover:enabled {
    background-color: #004f55;
    color: #fff;
}

.SmartNodeSelector input {
    border: none;
    background: none;
    outline: none;
    padding: 0px;
    width: 50px;
    color: #565656;
    display: inline-block;
}

.SmartNodeSelector ul {
    margin: 0;
    padding: 0;
    list-style-type: none;
}

.SmartNodeSelector__Tags--nolinebreak {
    display: inline-flex;
    flex-wrap: wrap;
}

.SmartNodeSelector__Tag {
    display: flex;
    flex-wrap: wrap;
    border-radius: 4px;
    background-color: #fefefe;
    justify-content: center;
    align-items: center;
    min-width: 0;
    position: relative;
    margin-right: 8px;
    margin-top: 4px;
    margin-bottom: 4px;
    color: #565656;
    border: 2px transparent solid;
    position: relative;
    white-space: pre;
}

.SmartNodeSelector__InnerTag {
    display: flex;
    flex: 1;
    border-radius: 4px;
    background-color: #fefefe;
    white-space: nowrap;
    padding: 8px;
    color: #565656;
}

.SmartNodeSelector__InputContainer {
    display: flex;
    white-space: nowrap;
    position: relative;
}

.SmartNodeSelector__Border {
    border: 2px #ccc solid;
}

.SmartNodeSelector__Ruler {
    visibility: hidden;
    white-space: nowrap;
}

.SmartNodeSelector__TagSelected {
    background-color: #d5eaf4;
    opacity: 0.5;
    position: absolute;
    left: 0px;
    top: 0px;
    width: 100%;
    height: 100%;
    display: block;
    z-index: 10;
}

.SmartNodeSelector__NumberOfTags {
    align-items: right;
    display: block;
    text-align: right;
    position: relative;
    width: 100%;
    margin-top: 8px;
    color: #6f6f6f;
}

.SmartNodeSelector__Warning {
    color: #ff9200;
}

.SmartNodeSelector__Error {
    color: #eb0000;
}

.SmartNodeSelector__MatchCounter {
    align-items: center;
    background: rgb(245, 245, 245);
    border: 1px #ccc solid;
    border-radius: 11px;
    color: #565656;
    height: 20px;
    justify-content: center;
    margin-right: 8px;
    padding: 0;
    padding-left: 3px;
    padding-right: 3px;
    min-width: 20px;
    display: inline-block;
    outline: none;
    position: relative;
    text-align: center;
    line-height: 20px;
}

.SmartNodeSelector__RemoveButton {
    position: absolute;
    right: -9px;
    top: -9px;
    background: #007079;
    border: 1px #fff solid;
    border-radius: 50%;
    cursor: pointer;
    width: 18px;
    height: 18px;
    padding: 0;
    display: block;
    background-image: url("./images/x.svg");
    background-position: center center;
    background-size: 14px 14px;
    background-repeat: no-repeat;
}

.SmartNodeSelector__RemoveButton:hover {
    color: white;
    background-color: #004f55;
    border: 1px #fff solid;
}

.SmartNodeSelector__Icon {
    background-size: 20px 20px;
    background-repeat: no-repeat;
    background-position: 5px center;
    padding-left: 30px;
    text-transform: capitalize;
}

.SmartNodeSelector__FocusOverlay {
    border-bottom: 1px #007079 dashed;
    position: absolute;
    height: 2px;
    bottom: 0px;
}

.SmartNodeSelector__Invalid,
.SmartNodeSelector__MissingSubgroup {
    background-color: #ffc1c1 !important;
    border: 2px #eb0000 solid !important;
}

.SmartNodeSelector__Duplicate {
    background-color: #ffe7d6 !important;
    border: 2px #ff9200 solid !important;
}

.SmartNodeSelector__InnerInvalid,
.SmartNodeSelector__InnerMissingSubgroup {
    background-color: #ffc1c1 !important;
    background-image: url("./images/invalid.svg");
}

.SmartNodeSelector__InnerDuplicate {
    background-color: #ffe7d6 !important;
    background-image: url("./images/duplicate.svg");
}

.SmartNodeSelector__Unknown {
    background-image: url("./images/unknown.svg");
}
