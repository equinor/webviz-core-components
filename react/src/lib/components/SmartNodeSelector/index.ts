/**
 * Copyright (c) 2021- Equinor ASA
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import TreeData, { MatchType } from "./utils/TreeData";
import TreeNodeSelection from "./utils/TreeNodeSelection";
import { TreeDataNode, TreeDataNodeMetaData } from "./utils/TreeDataNodeTypes";
import { SmartNodeSelectorPropsType } from "./components/SmartNodeSelectorComponent";
import SmartNodeSelectorComponent from "./components/SmartNodeSelectorComponent";

export { default } from "./SmartNodeSelector";
export {
    SmartNodeSelectorComponent,
    TreeData,
    MatchType,
    TreeDataNode,
    TreeDataNodeMetaData,
    SmartNodeSelectorPropsType,
    TreeNodeSelection,
};
