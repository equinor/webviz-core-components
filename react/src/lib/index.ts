/**
 * Copyright (c) 2021- Equinor ASA
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import WebvizPluginPlaceholder from "./components/WebvizPluginPlaceholder";
import { ColorScales } from "./components/ColorScales";
import Select from "./components/Select";
import SmartNodeSelector from "./components/SmartNodeSelector";
import { Menu } from "./components/Menu";
import { Overlay } from "./components/Overlay";
import { ScrollArea } from "./components/ScrollArea";

import "./components/FlexBox/flexbox.css";
import "./components/Layout";

export {
    WebvizPluginPlaceholder,
    ColorScales,
    Select,
    SmartNodeSelector,
    Menu,
    Overlay,
    ScrollArea,
};
