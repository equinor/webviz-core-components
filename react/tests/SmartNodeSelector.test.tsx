import React from "react";
import { fireEvent, render, RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SmartNodeSelector } from "../src/lib";
import { TreeDataNode } from "../src/lib/components/SmartNodeSelector";
import { SmartNodeSelectorInteractiveContainer } from "./SmartNodeSelectorInteractiveContainer";

export type PropType = {
    selectedTags: string[];
    selectedNodes: string[];
    selectedIds: string[];
};

let parentProps: PropType = {
    selectedTags: [],
    selectedNodes: [],
    selectedIds: [],
};

const setProps = (props: PropType): void => {
    parentProps = props;
};

enum RenderDataStructure {
    Flat = 1,
    FlatTwoNodes,
    Deep,
    DeepWithMetaData,
    InvalidData,
}

const clearParentProps = () => {
    parentProps = {
        selectedTags: [],
        selectedNodes: [],
        selectedIds: [],
    };
};

const createDataStructure = (
    renderDataStructure: RenderDataStructure
): TreeDataNode[] => {
    let data = [];
    switch (renderDataStructure) {
        case RenderDataStructure.Flat:
            data = [
                {
                    id: "1",
                    name: "Data",
                    description: "Description",
                },
            ];
            break;
        case RenderDataStructure.FlatTwoNodes:
            data = [
                {
                    id: "1",
                    name: "Data1",
                    description: "Description",
                },
                {
                    id: "2",
                    name: "Data2",
                    description: "Description",
                },
            ];
            break;
        case RenderDataStructure.Deep:
            data = [
                {
                    id: "1",
                    name: "Data",
                    description: "Description",
                    children: [
                        {
                            id: "1.1",
                            name: "Subdata",
                            description: "Description",
                        },
                    ],
                },
            ];
            break;
        case RenderDataStructure.DeepWithMetaData:
            data = [
                {
                    id: "1",
                    name: "Metadata 1",
                    description: "A first data source",
                    color: "#0095FF",
                    children: [
                        {
                            id: "1.1.1",
                            name: "Node 1",
                            description: "A first data node",
                            children: [
                                {
                                    id: "1.1.1.1",
                                    name: "Subnode 1",
                                    description: "A first sub node",
                                },
                            ],
                        },
                    ],
                },
            ];
            break;
        case RenderDataStructure.InvalidData:
            data = [
                {
                    id: "1",
                    name: "Metadata 1",
                    description: "Description",
                    children: [
                        {
                            id: "1.1",
                            name: "",
                            description: "Description",
                        },
                    ],
                },
            ];
    }
    return data;
};

const renderSmartNodeSelector = (
    renderDataStructure: RenderDataStructure,
    options: {
        showSuggestions?: boolean;
        initialTags?: string[];
        maxNumSelectedNodes?: number;
    } = {
        showSuggestions: true,
        initialTags: [],
        maxNumSelectedNodes: -1,
    }
): RenderResult => {
    const data = createDataStructure(renderDataStructure);
    return render(
        <SmartNodeSelector
            id="SmartNodeSelector"
            key="SmartNodeSelector"
            delimiter=":"
            selectedTags={options.initialTags}
            showSuggestions={options.showSuggestions}
            setProps={setProps}
            label="Smart Node Selector"
            numSecondsUntilSuggestionsAreShown={0.5}
            data={data}
        />
    );
};

const renderInteractiveSmartNodeSelector = (
    renderDataStructure: RenderDataStructure
): RenderResult => {
    const data = createDataStructure(renderDataStructure);
    return render(
        <SmartNodeSelectorInteractiveContainer
            setProps={setProps}
            data={data}
        />
    );
};

describe("SmartNodeSelector", () => {
    afterEach(() => {
        clearParentProps();
    });

    it("Renders correctly (compare to snapshot in ./__snapshots__/SmartNodeSelector.test.tsx.snap)", () => {
        const { container } = renderSmartNodeSelector(RenderDataStructure.Flat);
        expect(container).toMatchSnapshot();
    });

    it("Entering valid data node name and checking if tag is created (tree levels: 1)", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Flat,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            firstTag.classList.contains("SmartNodeSelector__Border")
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();
    });

    it("Entering valid data node name and checking if tag is created (tree levels: 2)", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Deep,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data:");

        expect(
            (firstTag.children[1] as HTMLElement).classList.contains(
                "SmartNodeSelector__InnerTag"
            )
        ).toBeTruthy();
        expect(firstTag.title === "Invalid").toBeTruthy();

        userEvent.type(firstInput, "Subdata");

        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            firstTag.classList.contains("SmartNodeSelector__Border")
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();
    });

    it("Entering valid data node name and then a duplicate", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Flat,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        const secondTag = smartNodeSelector.querySelectorAll(
            ".SmartNodeSelector__Tag"
        )[1] as HTMLElement;
        const secondInput = secondTag.querySelector("input");

        userEvent.type(secondInput, "Data");
        fireEvent.keyDown(secondInput, { key: "Enter" });
        fireEvent.keyUp(secondInput, { key: "Enter" });

        expect(
            secondTag.classList.contains("SmartNodeSelector__Duplicate")
        ).toBeTruthy();
        expect(secondTag.title === "Duplicate").toBeTruthy();
    });

    it("Entering invalid data source name", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Deep,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data:Error");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            firstTag.classList.contains("SmartNodeSelector__Invalid")
        ).toBeTruthy();
        expect(firstTag.title === "Invalid").toBeTruthy();
    });

    /*
    This test is not working anymore since the suggestions popup was moved to body. Apparently, 
    RTL does not recognize this properly. Disabled the test until a method is find to implemented
    it again.

    - RMT

    it("Suggestions are shown and contain expected options", (done) => {
        // eslint-disable-next-line no-undef
        jest.useFakeTimers();
        const { container } = renderSmartNodeSelector(RenderDataStructure.Deep);

        const smartNodeSelector = container.firstChild as HTMLElement;
        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        fireEvent.mouseDown(firstInput);

        setTimeout(() => {
            const suggestions = document.body.querySelector(
                ".Suggestions:not(.Suggestions__Position)"
            );
            expect(suggestions !== null).toBeTruthy();
            if (suggestions !== null) {
                const firstSuggestion = suggestions.querySelector(
                    ".Suggestions__Suggestion"
                );
                expect(firstSuggestion !== null).toBeTruthy();
            }
            done();
        }, 500);
        // eslint-disable-next-line no-undef
        jest.advanceTimersByTime(500);
    });
    */

    it("Remove button is working", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Flat,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            (firstTag.children[1] as HTMLElement).classList.contains(
                "SmartNodeSelector__InnerTag"
            )
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();
        expect(
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                .length == 2
        ).toBeTruthy();

        const removeButton = firstTag.querySelector(
            ".SmartNodeSelector__RemoveButton"
        );
        expect(removeButton !== null).toBeTruthy();
        if (removeButton !== null) {
            userEvent.click(removeButton);
            expect(
                smartNodeSelector
                    .querySelector(".SmartNodeSelector__Tag")
                    .classList.contains("SmartNodeSelector__Border")
            ).toBeFalsy();
            expect(
                smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
            ).toHaveLength(1);
        }
    });

    it("Clear all button is working", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Flat,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            (firstTag.children[1] as HTMLElement).classList.contains(
                "SmartNodeSelector__InnerTag"
            )
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();
        expect(
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                .length == 2
        ).toBeTruthy();

        const clearAllButton = smartNodeSelector.querySelector(
            ".SmartNodeSelector__ClearAll"
        ).firstChild as HTMLDivElement;

        expect(clearAllButton !== null).toBeTruthy();
        if (clearAllButton !== null) {
            userEvent.click(clearAllButton);
            expect(
                smartNodeSelector
                    .querySelector(".SmartNodeSelector__Tag")
                    .classList.contains("SmartNodeSelector__Border")
            ).toBeFalsy();
            expect(
                smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
            ).toHaveLength(1);
        }
    });

    const selectingTagsWithKeyboard = (numTags: number): HTMLElement => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Flat,
            { showSuggestions: false }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        let lastTag: HTMLElement | undefined = undefined;
        let lastInput: HTMLInputElement | undefined = undefined;

        for (let i = 0; i < numTags; i++) {
            lastTag = smartNodeSelector.querySelectorAll(
                ".SmartNodeSelector__Tag"
            )[
                smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                    .length - 1
            ] as HTMLElement;
            lastInput = lastTag.querySelector("input");
            userEvent.type(lastInput, "Data");
            fireEvent.keyDown(lastInput, { key: "Enter" });
            fireEvent.keyUp(lastInput, { key: "Enter" });
        }

        lastTag = smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")[
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                .length - 1
        ] as HTMLElement;
        lastInput = lastTag.querySelector("input");

        userEvent.click(lastInput);

        for (let i = 0; i < numTags; i++) {
            fireEvent.keyDown(i == 0 ? lastInput : document, {
                key: "ArrowLeft",
                shiftKey: true,
            });
        }

        return smartNodeSelector;
    };

    it("Selecting several tags with keyboard and delete using backspace", () => {
        const smartNodeSelector = selectingTagsWithKeyboard(3);
        expect(
            smartNodeSelector.querySelectorAll(
                ".SmartNodeSelector__TagSelected"
            )
        ).toHaveLength(3);
        fireEvent.keyDown(document, { key: "Backspace" });
        expect(
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
        ).toHaveLength(1);
    });

    it("Selecting several tags with keyboard and delete using delete", () => {
        const smartNodeSelector = selectingTagsWithKeyboard(3);

        expect(
            smartNodeSelector.querySelectorAll(
                ".SmartNodeSelector__TagSelected"
            )
        ).toHaveLength(3);
        fireEvent.keyDown(document, { key: "Delete" });
        expect(
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
        ).toHaveLength(1);
    });

    it("Selecting several tags and copy and pasting them", () => {
        const smartNodeSelector = selectingTagsWithKeyboard(3);
        const lastTag = smartNodeSelector.querySelectorAll(
            ".SmartNodeSelector__Tag"
        )[
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                .length - 1
        ] as HTMLElement;
        const lastInput = lastTag.querySelector("input");

        expect(
            smartNodeSelector.querySelectorAll(
                ".SmartNodeSelector__TagSelected"
            )
        ).toHaveLength(3);
        fireEvent.keyDown(document, { key: "c", ctrlKey: true });
        userEvent.click(lastInput);
        fireEvent.keyDown(lastInput, { key: "v", ctrlKey: true });

        expect(
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
        ).toHaveLength(7);
    });

    it("Check if return values are correct", () => {
        const smartNodeSelector = selectingTagsWithKeyboard(3);
        const lastTag = smartNodeSelector.querySelectorAll(
            ".SmartNodeSelector__Tag"
        )[
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                .length - 1
        ] as HTMLElement;
        const lastInput = lastTag.querySelector("input");

        fireEvent.keyDown(document, { key: "c", ctrlKey: true });
        userEvent.click(lastInput);
        fireEvent.keyDown(lastInput, { key: "v", ctrlKey: true });

        expect(parentProps.selectedTags).toHaveLength(1);
        expect(parentProps.selectedTags).toEqual(["Data"]);
        expect(parentProps.selectedNodes).toHaveLength(1);
        expect(parentProps.selectedNodes[0]).toMatch("Data");
        expect(parentProps.selectedIds).toHaveLength(1);
        expect(parentProps.selectedIds[0]).toMatch("1");

        clearParentProps();
    });

    it("Check if initial properties can be set", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Deep,
            {
                showSuggestions: false,
                initialTags: ["Data:Subdata"],
            }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        expect(
            (firstTag.children[1] as HTMLElement).classList.contains(
                "SmartNodeSelector__InnerTag"
            )
        ).toBeTruthy();
        expect(firstTag.title === "Data:Subdata").toBeTruthy();
        expect(
            smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
                .length == 2
        ).toBeTruthy();
    });

    it("Check if selected tags can be dynamically changed", () => {
        const { container } = renderInteractiveSmartNodeSelector(
            RenderDataStructure.FlatTwoNodes
        );

        const smartNodeSelector = container.querySelector("#SmartNodeSelector");
        expect(smartNodeSelector).toBeDefined();

        const button = container.querySelector("#setValuesButton");
        expect(button).toBeDefined();

        expect(parentProps.selectedTags).toHaveLength(0);
        expect(parentProps.selectedNodes).toHaveLength(0);

        userEvent.click(button);

        expect(parentProps.selectedTags).toHaveLength(2);
        expect(parentProps.selectedTags[0]).toMatch("Data1");
        expect(parentProps.selectedTags[1]).toMatch("Data2");
        expect(parentProps.selectedNodes).toHaveLength(2);
        expect(parentProps.selectedNodes[0]).toMatch("Data1");
        expect(parentProps.selectedNodes[1]).toMatch("Data2");
        expect(parentProps.selectedIds).toHaveLength(2);
        expect(parentProps.selectedIds[0]).toMatch("1");
        expect(parentProps.selectedIds[1]).toMatch("2");

        clearParentProps();
    });

    it("Check if data can be dynamically changed", () => {
        const { container } = renderInteractiveSmartNodeSelector(
            RenderDataStructure.FlatTwoNodes
        );

        const smartNodeSelector = container.querySelector("#SmartNodeSelector");
        expect(smartNodeSelector).toBeDefined();

        const button = container.querySelector("#setDataButton");
        expect(button).toBeDefined();

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data1");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            firstTag.classList.contains("SmartNodeSelector__Border")
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();

        userEvent.click(button);

        expect(
            firstTag.classList.contains("SmartNodeSelector__Invalid")
        ).toBeTruthy();
        expect(firstTag.title === "Invalid").toBeTruthy();

        const clearAllButton = smartNodeSelector.querySelector(
            ".SmartNodeSelector__ClearAll"
        ).firstChild as HTMLDivElement;

        expect(clearAllButton !== null).toBeTruthy();
        if (clearAllButton !== null) {
            userEvent.click(clearAllButton);
            expect(
                smartNodeSelector
                    .querySelector(".SmartNodeSelector__Tag")
                    .classList.contains("SmartNodeSelector__Border")
            ).toBeFalsy();
            expect(
                smartNodeSelector.querySelectorAll(".SmartNodeSelector__Tag")
            ).toHaveLength(1);
        }

        userEvent.type(firstInput, "ChangedData1");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            firstTag.classList.contains("SmartNodeSelector__Border")
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();

        clearParentProps();
    });

    it("Check that SmartNodeSelector is working well when used as a controlled component", () => {
        const { container } = renderInteractiveSmartNodeSelector(
            RenderDataStructure.Deep
        );

        const smartNodeSelector = container.querySelector("#SmartNodeSelector");
        expect(smartNodeSelector).toBeDefined();

        const firstInput = smartNodeSelector.querySelector(
            "input"
        ) as HTMLInputElement;

        userEvent.type(firstInput, "Data");
        fireEvent.keyDown(firstInput, { key: ":" });
        fireEvent.keyUp(firstInput, { key: ":" });

        expect(firstInput.value === "Data:").toBeTruthy();

        clearParentProps();
    });

    it("Check if selecting a node in a single node selection is calling setProps correctly", () => {
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.Flat,
            {
                showSuggestions: false,
                maxNumSelectedNodes: 1,
            }
        );

        const smartNodeSelector = container.firstChild as HTMLElement;

        const firstTag = smartNodeSelector.querySelector(
            ".SmartNodeSelector__Tag"
        ) as HTMLElement;
        const firstInput = firstTag.querySelector("input");

        userEvent.type(firstInput, "Data");
        fireEvent.keyDown(firstInput, { key: "Enter" });
        fireEvent.keyUp(firstInput, { key: "Enter" });

        expect(
            firstTag.classList.contains("SmartNodeSelector__Border")
        ).toBeTruthy();
        expect(firstTag.title === firstInput.value).toBeTruthy();

        expect(parentProps.selectedTags).toHaveLength(1);
        expect(parentProps.selectedTags[0]).toMatch("Data");
        expect(parentProps.selectedNodes).toHaveLength(1);
        expect(parentProps.selectedNodes[0]).toMatch("Data");
        expect(parentProps.selectedIds).toHaveLength(1);
        expect(parentProps.selectedIds[0]).toMatch("1");

        clearParentProps();
    });

    it("Ensure that invalid data throws exception", () => {
        jest.spyOn(console, "error").mockImplementation(jest.fn());
        const { container } = renderSmartNodeSelector(
            RenderDataStructure.InvalidData
        );

        const smartNodeSelector = container.firstChild as HTMLElement;
        expect(
            smartNodeSelector.classList.contains("SmartNodeSelector--Error")
        ).toBeTruthy();
        jest.spyOn(console, "error").mockRestore();
    });
});
