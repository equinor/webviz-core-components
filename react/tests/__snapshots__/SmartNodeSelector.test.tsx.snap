// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartNodeSelector Renders correctly (compare to snapshot in ./__snapshots__/SmartNodeSelector.test.tsx.snap) 1`] = `
<div>
  <div
    id="SmartNodeSelector"
  >
    <label>
      Smart Node Selector
    </label>
    <div
      class="SmartNodeSelector"
    >
      <ul
        class="SmartNodeSelector__Tags--nolinebreak"
      >
        <li
          class="SmartNodeSelector__Tag "
          title="Enter a new name"
        >
          <div
            class="SmartNodeSelector__InnerTag"
          >
            <div
              class="SmartNodeSelector__InputContainer"
            >
              <input
                placeholder="Add new tag..."
                spellcheck="false"
                style="width: 100px;"
                type="text"
                value=""
              />
            </div>
          </div>
        </li>
      </ul>
      <div
        class="SmartNodeSelector__ClearAll"
      >
        <button
          disabled=""
          title="Clear all"
          type="button"
        />
      </div>
      <div
        class="Suggestions"
        style="max-height: 568px; display: none;"
      >
        <div
          class="Suggestions__Spacer"
          style="height: 0px;"
        />
        <div
          class="Suggestions__Suggestion Suggestions__Suggestion--Selected"
          data-index="0"
          data-use="Data"
          style="background-image: none; height: 34px;"
        >
          <span
            class="Suggestions__Match"
          />
          Data
           - Description
        </div>
        <div
          class="Suggestions__Spacer"
          style="height: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;
